endpoint/verify.go

```diff
@@ -0,0 +1,100 @@
+package endpoint
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/google/uuid"
+)
+
+// Authorization is used to interact with the `jwt` package
+type Authorization struct {
+	ID      string `json:"sub"`
+	Session string `json:"https://ataper.net/session"`
+	Role    string `json:"https://ataper.net/role"`
+
+	Phone string `json:"https://ataper.net/phone"`
+	Email string `json:"https://ataper.net/email"`
+
+	Issuer string `json:"iss"`
+}
+
+// Anonymous returns whether this authorization lacks an identity
+func (a Authorization) Anonymous() bool {
+	return a.ID == ""
+}
+
+// Valid implements jwt.Authorization
+func (a Authorization) Valid() error {
+	return nil
+}
+
+// NewVerifyEndpoint returns a new HTTP handler for requests to the
+// /v1/verify endpoint, which is used for forward-auth with traefik.
+func NewVerifyEndpoint(
+	serviceDomain string,
+	jwtSigningSecret string,
+) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		authHeader, ok := r.Header["Authorization"]
+		if !ok || len(authHeader) == 0 {
+			tokenString, err := jwt.NewWithClaims(
+				jwt.GetSigningMethod(jwt.SigningMethodHS256.Alg()),
+				generateAnonymousAuthorization(serviceDomain),
+			).SignedString([]byte(jwtSigningSecret))
+			if err != nil {
+				fmt.Println("[ERROR]", err.Error())
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusForbidden)
```

Comment:
why does not having an auth header return a 403 but having a malformated header return a 401. I feel like they should both be 401

Reply:
Good call - in both cases we're complaining that there is no valid authorization at all, there's no "missing permissions" here.